### ISSUE TITLE
Azure CNI Powered by Cilium L7/CRD Network Policy

### ISSUE THREAD

**Issue by chasewilson** [Azure/AKS#3797]
## GA ETA: mid-November, 2025*
*ETAs are estimations and subject to change.

This issue is to track support for Azure CNI Powered by Cilium Network Policy capability expansions. 

Expansions Include: 
1. Layer 7 Network Policy support
2. Cilium Custom Network Policies
    - Currently, one of the most requisition options is DNS filtering based policy support 

This feature has a some good support already tracked in [this GitHub Issue](https://github.com/Azure/AKS/issues/3450).

**Comment by Hanifff** [comment]
Hi, any updates?

**Comment by chasewilson** [comment]
Hey @Hanifff, thanks for commenting here. This Item is to track the request and interest for these features in Azure CNI Powered by Cilium. Right now there isn't a timeline but we are keeping an eye on what our customers want and would like it to build as needs arise.

**Comment by westleydion** [comment]
This would be really good to have.

**Comment by chasewilson** [comment]
> This would be really good to have.

Which portion of this are you most interested in? The Cilium Specific policies or the L7 capabilities?

**Comment by ebc92** [comment]
My team is also missing this feature, DNS based network policies is a must have for us.

**Comment by illrill** [comment]
Not stale. Lack of CiliumNetworkPolicy L3 FQDN rules is one of the reasons why we still need to BYOCNI, just to even use the most basic features of Cilium.

**Comment by EppO** [comment]
The issue is still relevant

**Comment by lieberlois** [comment]
Apparently, Azure will start supporting this in 1-2 months. Source: Talked to Isovalent employees at the KCD in Munich, Germany. Unfortunatley, Hubble UI integration etc. will take longer.

**Comment by chasewilson** [comment]
@lieberlois thanks for the input here. Would you mind clarifying in what scenario Isovalent was planning L7 support? 

From our side, we currently have support for [hubble relay with self-managed UI](https://learn.microsoft.com/en-us/azure/aks/advanced-container-networking-services-overview) and we're not working on l7 quite yet but are working on supporting [FQDN filtering](https://github.com/Azure/AKS/issues/4205) hopefully by the end of this month.

**Comment by lieberlois** [comment]
@chasewilson This was in the context of layer 7 network policies 😄

**Comment by chasewilson** [comment]
@lieberlois sorry for the confusion 😆. 

I was meaning, did they say specifically [Azure CNI Powered by Cilum](https://learn.microsoft.com/en-us/azure/aks/azure-cni-powered-by-cilium), The [enterprise marketplace offering](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/isovalentinc1662143158090.isovalent-cilium-enterprise?tab=overview) they have, or the OSS Cilium support?

**Comment by lieberlois** [comment]
@chasewilson As far as I understood yes, [Azure CNI Powered by Cilium](https://learn.microsoft.com/en-us/azure/aks/azure-cni-powered-by-cilium)

**Comment by chasewilson** [comment]
@lieberlois thanks for the clarification. 

As of right now, we're not on L7 yet as we've had more requests (though L7 is highly requested as well) for FQDN and will be aiming for L7 after we get that out. So, not in the next month or two but should have some updates on timelines within that period.

**Comment by lieberlois** [comment]
What exactly are you referring to then? Layer 7 network policies leverage FQDNs so what is missing then?

**Comment by chasewilson** [comment]
> What exactly are you referring to then? Layer 7 network policies leverage FQDNs so what is missing then?

Good question. 

Cilium L7 policies and FQDN policies both work at Layer 7, but they have different focuses. L7 policies give you detailed control over app-specific traffic, letting you set rules based on things like HTTP methods or gRPC services. 

On the other hand, FQDN policies are about controlling outbound traffic based on domain names. This is helpful in dynamic environments where IP addresses of external services change, but domain names stay the same.

**Comment by lieberlois** [comment]
Okay you seem to have different naming than the typical Service Mesh terminology then 😄 I meant egress policies based on L7 Hostnames (FQDNs)

**Comment by chasewilson** [comment]
@lieberlois aaahhh gotcha ok and I'm referring to L7 as application operations traffic. PUTs, GETs, etc.

**Comment by siegenthalerroger** [comment]
Still relevant

**Comment by laurenbo** [comment]
still relevant also, we await eagerly for a public preview with the FQDN policies.

**Comment by maur1** [comment]
still relevant also, we await eagerly for a public preview with the FQDN policies

**Comment by EvertonSA** [comment]
still relevant

**Comment by sebnyberg** [comment]
Will L7 support will be added to the Advanced Container Network Services (ACNS) offering?

(Also keenly awaiting this feature! It's truly a game changer.)

**Comment by quantumn-a5** [comment]
Yes, and its live now ( in Public Preview). Learn more here : [Introducing Layer 7 Network Policies with Advanced Container Networking Services for AKS Clusters! | Microsoft Community Hub](https://techcommunity.microsoft.com/blog/azurenetworkingblog/introducing-layer-7-network-policies-with-advanced-container-networking-services/4409782).

**Comment by geowalrus4gh** [comment]
is this feature available only for ACNS enabled users?

**Comment by quantumn-a5** [comment]
@geowalrus4gh - Yes, that's correct.

**Comment by danbosscher** [comment]
Hi folks - just letting you know we aim to promote this to GA in mid-November, 2025.

**Comment by wayden88** [comment]
@danbosscher and @chasewilson,

Hi team,

Quick question—I have a customer who deployed Datadog to their AKS clusters using a Helm chart. Everything was working fine with Cilium Network Policy until yesterday (Nov 8, 2025). After restarting Datadog, the cluster is now rejecting the Datadog Cilium Network Policy with the following error:
**Failed sync attempt to: one or more objects failed to apply, reason: ciliumnetworkpolicies.cilium.io 'datadog' is forbidden: ValidatingAdmissionPolicy 'advanced-networking-validating-policy' with binding 'advanced-networking-cilium-l7-binding' denied request: CiliumNetworkPolicy cannot use DNS L7 rules.**

Is this related to a known issue (same as this this one) or any recent change in policy enforcement? The customer mentioned it was working previously without any issues. Is there a fix expected around mid-November 2025?

**Comment by danbosscher** [comment]
@shaifaligargmsft if you can have a look?
+ https://github.com/Azure/AKS/issues/4663

**Comment by shaifaligargmsft** [comment]
> [@shaifaligargmsft](https://github.com/shaifaligargmsft) if you can have a look?
> 
> * [[BUG] `CiliumNetworkPolicy` is invalid: `ValidatingAdmissionPolicy` 'advanced-networking-validating-policy' with binding 'advanced-networking-cilium-l7-binding' denied request #4663](https://github.com/Azure/AKS/issues/4663)

@nddq has responded that with the given network profile configuration, its not clear if they enabled L7 policy support along with ACNS. he has advised to update azure cli so that API model have one specific parameter "advancedNetworkPolicies" field in the networkProfile . this parameter can assure if customer has enabled L7 policy support along with ACNS.

**Comment by n-gras** [comment]
@danbosscher and @chasewilson,

Hi team,

I've enabled ACNS with Cilium Network Policies and L7 filtering using TLS inspection. However, adding even a simple 
```
rules: 
  http: 
    - {} 
```
Breaks websockets connections to our cluster (403 responses on websocket requests), while other traffic stays working as expected. As soon as I remove the L7 policy part, the websocket traffic works again.

Might it be related to this issue? (https://github.com/cilium/cilium/issues/11672), which means it should already be solved upstream.

**Comment by nddq** [comment]
@n-gras looks like the issue have been resolved in Cilium v1.18.3, which we are planning to release soon to couple with k8s v1.34 GA

**Comment by nddq** [comment]
@danbosscher I've worked with @wayden88 to resolve the issue you've linked. It was due to a missing field in the cluster's state that got reconciled after doing a blank `az aks update` with the latest API version.

**Comment by n-gras** [comment]
@nddq that's great to hear, thank you for the update!

**Comment by sf-msft** [comment]
Closing as completed. Terraform via azure-rm support is in-flight. Please chime in over at https://github.com/hashicorp/terraform-provider-azurerm/pull/31506 for any follow up.
